src/task_processor/threads.py

```diff
@@ -95,18 +96,46 @@ def run(self) -> None:
             time.sleep(self.sleep_interval_millis / 1000)
 
     def run_iteration(self) -> None:
-        try:
-            run_tasks(self.queue_pop_size)
-            run_recurring_tasks()
-        except Exception as e:
-            # To prevent task threads from dying if they get an error retrieving the tasks from the
-            # database this will allow the thread to continue trying to retrieve tasks if it can
-            # successfully re-establish a connection to the database.
-            # TODO: is this also what is causing tasks to get stuck as locked? Can we unlock
-            #  tasks here?
-
-            logger.error("Received error retrieving tasks: %s.", e, exc_info=e)
-            close_old_connections()
+        """
+        Consume and execute tasks from the queue, and run recurring tasks
+
+        This method tries to consume tasks from multiple databases as to ensure
+        that any remaining tasks are processed after opting in or out of a
+        separate database setup.
+        """
+        database_is_separate = "task_processor" in settings.TASK_PROCESSOR_DATABASES
+        if database_is_separate:
+            assert (
+                "task_processor.routers.TaskProcessorRouter"
+                in settings.DATABASE_ROUTERS
```

Comment:
We won't have to do this check as well if we only include the router if `task_processor` exists. 
ps: related to the other comment

Reply:
This is a safeguard to ensure whatever happens with `DATABASE_ROUTERS` when [flagsmith settings](https://github.com/Flagsmith/flagsmith/blob/390ffc8395022e136e8bf2a77139447b3d5435b6/api/app/settings/common.py) are evaluated won't break the task processor.

Relating to the other comment, if we decide against checking `settings.TASK_PROCESSOR_DATABASES`, I'd still keep this guard-railed, e.g:

```python
assert "task_processor" in settings.DATABASES and "task_processor.routers.TaskProcessorRouter" in settings.DATABASE_ROUTERS
```

Note that this `assert` only evaluates when either 1. Flagsmith default settings (see above) or 2. the user has explicitly decided to include `"task_processor"` in `settings.TASK_PROCESSOR_DATABASES`.

Reply:
I like it. Fixed in 3fef27327d5f36ce439b129f23746f9ca65da3b2.